### ISSUE TITLE
Batch vs layer normalization description (page 560)

### ISSUE THREAD

**Issue by labdmitriy** [rasbt/machine-learning-book#40]
Hi Sebastian,
 
There is the description of the batch and layer normalization (including the picture) on the page 560:
 
"While **layer** normalization is traditionally performed across all elements in a given feature for each feature independently, the **layer** normalization used in transformers extends this concept and computes the normalization statistics across all feature values independently for each training example."
 
Is layer normalization mentioned correctly in the first case? It seems that when we calculate statistics for each feature independently, we perform batch normalization?
 
Thank you.

**Comment by rasbt** [comment]
Thanks for the note! Phew this is a tricky one

Originally:

![Screen Shot 2022-04-14 at 11 01 18 PM](https://user-images.githubusercontent.com/5618407/163516199-1c1f7f0b-1ed0-4e68-9673-6996b4194928.png)

> While layer normalization is traditionally performed across all elements in a given feature for each feature independently, the layer normalization used in transformers extends this concept and computes the normalization statistics across all feature values independently for each training example.



Yeah, reading this again, it does sound a bit weird. Here is an attempt to clarify it:

> While layer normalization is traditionally performed across all feature values in a given layer for each training example independently, the layer normalization used in transformers extends this concept and computes the normalization statistics across all feature values for a given sentence token position independently for each training example


Maybe I should also swap the figure with a better one. This is the figure from the original [layer norm paper](https://arxiv.org/abs/1607.06450v1):

<img width="452" alt="Screen_Shot_2020-05-19_at_4 24 42_PM" src="https://user-images.githubusercontent.com/5618407/163516319-97e2a11a-85e2-48e7-bd37-e0f90c08ceb7.png">


 I just found this helpful one [here](https://i.stack.imgur.com/E3104.png) for transformer contexts: 
<img width="307" alt="E3104-1" src="https://user-images.githubusercontent.com/5618407/163516379-ade1e98c-8f01-43ab-88d7-caa059077c4c.png">

Do you think that showing it like this would make it more clear? We can potentially update the book and I could swap it out. Do you think it's a helpful change?

**Comment by labdmitriy** [comment]
I don't have much practical experience in DL, but I suppose the following:
- The first picture has dimension annotations which are more specific to CNN than to sequences.
And it can be confusing to read PyTorch documentation, where for example convolutional layer will have N, C, L (for 1d) or N, C, H, W (for 2d) dimensions, while for embedding we will have swapped dimensions (N, L, C), and others can have (L, N, C) dimensions:
https://discuss.pytorch.org/t/inconsistent-dimension-ordering-for-1d-networks-ncl-vs-nlc-vs-lnc/14807
- The implementations which I saw for transformers, including ["The Annotated transformer"](https://nlp.seas.harvard.edu/2018/04/03/attention.html), as I understand, uses LayerNorm with normalization only by the last dimension (embeddings dimension), so it seems that classical Layer normalization is used.
- I suppose that the second picture is from this [article](https://arxiv.org/pdf/2003.07845.pdf), if you mean the left case in the picture then it is probably better for understanding your phrase. As I understand, this type of layer normalization is not described in the article in detail, just for reference.
- Unfortunately I didn't find any description of this extended type of normalization neither in articles (for example, [here](https://theaisummer.com/normalization/#layer-normalization-2016)) nor in implementations, and it will be great if you suggest any additional information about that.
